internal/api/terminal_manager.go

```diff
@@ -0,0 +1,174 @@
+	config config.Dash
+	poller Poller
+
+	sendScrollback sync.Map
+}
+
+type terminalOutput struct {
+	Scrollback  []byte `json:"scrollback,omitempty"`
+	Line        []byte `json:"line,omitempty"`
+	ExitMessage []byte `json:"exitMessage,omitempty"`
+}
+
+var _ StateManager = (*terminalStateManager)(nil)
+
+// NewTerminalStateManager returns a terminal state manager.
+func NewTerminalStateManager(cfg config.Dash) StateManager {
+	return &terminalStateManager{
+		config: cfg,
+		poller: NewInterruptiblePoller("terminal"),
+	}
+}
+
+// Handlers returns a slice of handlers.
+func (s *terminalStateManager) Handlers() []octant.ClientRequestHandler {
+	return []octant.ClientRequestHandler{
+		{
+			RequestType: RequestTerminalScrollback,
+			Handler:     s.SendTerminalScrollback,
+		},
+		{
+			RequestType: RequestTerminalCommand,
+			Handler:     s.SendTerminalCommand,
+		},
+		{
+			RequestType: RequestTerminalResize,
+			Handler:     s.SendTerminalResize,
+		},
+	}
+}
+
+func (s *terminalStateManager) SendTerminalResize(state octant.State, payload action.Payload) error {
+	terminalID, err := payload.String("terminalID")
+	if err != nil {
+		return errors.Wrap(err, "extract terminal ID from payload")
+	}
+
+	rows, err := payload.Uint16("rows")
+	if err != nil {
+		return errors.Wrap(err, "extract rows from payload")
+	}
+
+	cols, err := payload.Uint16("cols")
+	if err != nil {
+		return errors.Wrap(err, "extract cols from payload")
+	}
+
+	tm := s.config.TerminalManager()
+	t, ok := tm.Get(terminalID)
+	if !ok {
+		return errors.New(fmt.Sprintf("terminal %s not found", terminalID))
+	}
+
+	if t.Active() {
+		t.Resize(cols, rows)
+	}
+	return nil
+}
+
+func (s *terminalStateManager) SendTerminalCommand(state octant.State, payload action.Payload) error {
+	terminalID, err := payload.String("terminalID")
+	if err != nil {
+		return errors.Wrap(err, "extract terminal ID from payload")
+	}
+
+	key, err := payload.String("key")
+	if err != nil {
+		return errors.Wrap(err, "extract key from payload")
+	}
+
+	tm := s.config.TerminalManager()
+	t, ok := tm.Get(terminalID)
+	if !ok {
+		return errors.New(fmt.Sprintf("terminal %s not found", terminalID))
+	}
+	return t.Write([]byte(key))
+}
+
+func (s *terminalStateManager) SendTerminalScrollback(state octant.State, payload action.Payload) error {
+	terminalID, err := payload.String("terminalID")
+	if err != nil {
+		return errors.Wrap(err, "extract terminal ID from payload")
+	}
+	s.setSendScrollback(terminalID, true)
+	return nil
+}
+
+func (s *terminalStateManager) setSendScrollback(id string, v bool) {
+	s.sendScrollback.Store(id, v)
+}
+
+func (s *terminalStateManager) Start(ctx context.Context, state octant.State, client OctantClient) {
+	ch := make(chan struct{}, 1)
+	defer func() {
+		close(ch)
+	}()
+	s.poller.Run(ctx, ch, s.runUpdate(state, client), event.TerminalStreamDelay)
+}
+
+func (s *terminalStateManager) runUpdate(state octant.State, client OctantClient) PollerFunc {
```

Comment:
I think using a single poller to loop over terminals to send the websocket message is not great. I'd rather see this done as an individual websocket endpoint for each terminal.

I imagine this happening in the shape of a websocket broker that can register wss endpoints on demand.